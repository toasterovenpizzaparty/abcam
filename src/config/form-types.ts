// These steps enums serve as keys for our step object within our state.
export enum STEPS {
  STEP_ONE = "step-one",
  STEP_TWO = "step-two",
}

// These fields enums serve as keys for our fields within our step object.
export enum FIELDS {
  RATING = "rating",
  TITLE = "title",
  DESCRIPTION = "description",
  IMAGE = "image",
  IMAGE_DESCRIPTION = "image-description",
}
