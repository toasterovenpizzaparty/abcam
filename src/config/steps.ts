import React from "react";
import { StepDataType } from "../redux/reducers/step-reducer/step-reducer";
import { StepPropTypes } from "../types/types";
import { FIELDS } from "./form-types";

const StepOne = React.lazy(() => import("../pages/steps/step-one/step-one"));
const StepTwo = React.lazy(() => import("../pages/steps/step-two/step-two"));
const StepThree = React.lazy(
  () => import("../pages/steps/step-three/step-three")
);
const StepFour = React.lazy(() => import("../pages/steps/step-four/step-four"));

// Create a simple map to connect our step index to a component
export const StepComponentMap: Readonly<
  Record<number, () => React.FC<StepPropTypes>>
> = {
  0: () => StepOne,
  1: () => StepTwo,
  2: () => StepThree,
  3: () => StepFour,
};

export const StepDescriptionMap: Readonly<Record<number, string>> = {
  0: "Title and rating",
  1: "Description",
  2: "Image and description",
  3: "See your review",
};
export const numOfSteps = Object.keys(StepComponentMap).length;

export const getStepComponent = (index: number) =>
  StepComponentMap[index] || StepOne;

export const getStepIndex = (stepIndex: string | undefined) =>
  Math.max(0, (parseInt(stepIndex || "0") || 0) - 1);

export const getField = (state: StepDataType, field: string) =>
  (state?.fields && state?.fields[field]) || "";

export const getNextPageIndex = (index: number) =>
  Math.min(numOfSteps, index + 1);

export const getPreviousPageIndex = (index: number) => Math.max(0, index - 1);

export const getFields = (fields: string[], stepState: StepDataType) => {
  const mappedFields = fields.map((key) => {
    return {
      key,
      value: (stepState?.fields && stepState?.fields[key]) || null,
    };
  });
  if (mappedFields.length) {
    return mappedFields.reduce(
      (acc, field) => ({
        ...acc,
        ...(field.value !== null ? { [field.key]: field.value } : {}),
      }),
      {}
    );
  } else {
    return {};
  }
};

export const getButtonDisabled = (
  stepState: StepDataType,
  componentIndex: number
) => {
  switch (componentIndex) {
    case 0:
      return (
        [FIELDS.TITLE, FIELDS.RATING].filter(
          (field) => !getField(stepState, field)
        ).length > 0
      );
    case 2:
      return (
        [FIELDS.IMAGE].filter((field) => !getField(stepState, field)).length > 0
      );
    default:
      return false;
  }
};
