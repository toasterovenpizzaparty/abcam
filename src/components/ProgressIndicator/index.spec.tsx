import React from "react";
import { shallow } from "enzyme";
import "../../config/enzyme";
import { ProgressIndicator, getFloat, getPercentage } from "./index";

describe("<ProgressIndicator />", () => {
  it("Renders the progress indicator", () => {
    const wrapper = shallow(<ProgressIndicator current={25} />);
    expect(wrapper).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("Converts input to a proper float number", () => {
    // Verify string conversion
    expect(getFloat("0")).toEqual(0);
    // Verify string conversion from non-numeric value
    expect(getFloat("foobar")).toEqual(0.0);
    // Verify number conversion
    expect(getFloat(123)).toEqual(123);
    // Verify precision
    expect(getFloat(123.2343, 0.0, 2)).toEqual(123.23);
  });

  it("Converts input to a proper percentage number", () => {
    // Division by zero
    expect(getPercentage("0", "0")).toEqual(0);
    expect(getPercentage("0", "100")).toEqual(0);
    expect(getPercentage(25, 100)).toEqual(0.25);
    expect(getPercentage("25", "100")).toEqual(0.25);
  });
});
