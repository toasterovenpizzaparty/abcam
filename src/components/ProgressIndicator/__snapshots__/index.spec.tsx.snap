// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressIndicator /> Renders the progress indicator 1`] = `ShallowWrapper {}`;
